packages/lit-html/src/lib/directives/ref.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * @license
+ * Copyright (c) 2020 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Directive, directive, nothing, SpreadPart} from '../lit-html.js';
+
+/**
+ * A container for a reference to an element.
+ */
+export class Ref<T = Element> {
+  value?: T;
+
+  set(v: T): void {
+    this.value = v;
+  }
+}
+
+class RefDirective extends Directive {
+
+  render(_ref: Ref) {
+    return nothing;
+  }
+
+  update(part: SpreadPart, [ref]: Parameters<this['render']>) {
+    if (ref.value !== part.__element) {
+      ref.set(part.__element);
```

Comment:
I think that makes common usage a little more cumbersome. You'll always need a way to read a ref, so you need some container.

For a common case of linking two elements:

```ts
const fooRef = new Ref();
render(html`
  <x-foo ${ref(fooRef)}></x-foo>
  <x-bar .foo=${fooRef.value}></x-bar>
`, container);
```

you'd have to do:
```ts
const foo;
const fooRef (e) => fooRefValue = e;
render(html`
  <x-foo ${ref(fooRef)}></x-foo>
  <x-bar .foo=${foo}></x-bar>
`, container);
```

Not hugely different, but enough I think. I'm also looking forward towards possibly having a RefController that can trigger a re-render if the element changes. This would have the same `set()`/`value` API. React can take both Ref objects and callbacks, btw. We can discuss.